<a href="https://colab.research.google.com/github/Lasttrader/LT_Machine-learning/blob/master/ML_wine_colab_kNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Задача множественной классификации sklearn (k NN)

In [0]:
'''
План:
1. Загрузка данных
2. Провекра данных
3. Очистка
4. Присвоение Х и У (значения метки)
5. Создаем модель, определяем обучающую и проверочную выборки
6. Создаем модель
7. ОБучаем
8. Проверяем
9. Подбираем количество соседей и определяем наилучую точность

'''

# Импорт базы данных вина
from sklearn.datasets import load_wine

data_wine = load_wine()

# задача множественной классификации - вина
type(data_wine)

sklearn.utils.Bunch

In [0]:
# Классы вина
print(data_wine.target_names)
print(data_wine.feature_names)


['class_0' 'class_1' 'class_2']
['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']


In [0]:
# Выводим дата сет данные
print(data_wine.target)
print(list(data_wine.target_names))
print(data_wine.data.shape)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2]
['class_0', 'class_1', 'class_2']
(178, 13)


In [0]:

X = data_wine.data # значения

y = data_wine.target # классы

X.shape


(178, 13)

In [0]:
X[:5]

array([[1.423e+01, 1.710e+00, 2.430e+00, 1.560e+01, 1.270e+02, 2.800e+00,
        3.060e+00, 2.800e-01, 2.290e+00, 5.640e+00, 1.040e+00, 3.920e+00,
        1.065e+03],
       [1.320e+01, 1.780e+00, 2.140e+00, 1.120e+01, 1.000e+02, 2.650e+00,
        2.760e+00, 2.600e-01, 1.280e+00, 4.380e+00, 1.050e+00, 3.400e+00,
        1.050e+03],
       [1.316e+01, 2.360e+00, 2.670e+00, 1.860e+01, 1.010e+02, 2.800e+00,
        3.240e+00, 3.000e-01, 2.810e+00, 5.680e+00, 1.030e+00, 3.170e+00,
        1.185e+03],
       [1.437e+01, 1.950e+00, 2.500e+00, 1.680e+01, 1.130e+02, 3.850e+00,
        3.490e+00, 2.400e-01, 2.180e+00, 7.800e+00, 8.600e-01, 3.450e+00,
        1.480e+03],
       [1.324e+01, 2.590e+00, 2.870e+00, 2.100e+01, 1.180e+02, 2.800e+00,
        2.690e+00, 3.900e-01, 1.820e+00, 4.320e+00, 1.040e+00, 2.930e+00,
        7.350e+02]])

In [0]:
y[:5]

array([0, 0, 0, 0, 0])

In [0]:
# модель

from sklearn.model_selection import train_test_split

# Создаем разбивку и перемешку
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [0]:
# размерность множеств
print(X_train.shape, y_train.shape) # обучение
print(X_test.shape, y_test.shape) # проверка

(142, 13) (142,)
(36, 13) (36,)


In [0]:
X_train[:5]


array([[1.434e+01, 1.680e+00, 2.700e+00, 2.500e+01, 9.800e+01, 2.800e+00,
        1.310e+00, 5.300e-01, 2.700e+00, 1.300e+01, 5.700e-01, 1.960e+00,
        6.600e+02],
       [1.253e+01, 5.510e+00, 2.640e+00, 2.500e+01, 9.600e+01, 1.790e+00,
        6.000e-01, 6.300e-01, 1.100e+00, 5.000e+00, 8.200e-01, 1.690e+00,
        5.150e+02],
       [1.237e+01, 1.070e+00, 2.100e+00, 1.850e+01, 8.800e+01, 3.520e+00,
        3.750e+00, 2.400e-01, 1.950e+00, 4.500e+00, 1.040e+00, 2.770e+00,
        6.600e+02],
       [1.348e+01, 1.670e+00, 2.640e+00, 2.250e+01, 8.900e+01, 2.600e+00,
        1.100e+00, 5.200e-01, 2.290e+00, 1.175e+01, 5.700e-01, 1.780e+00,
        6.200e+02],
       [1.307e+01, 1.500e+00, 2.100e+00, 1.550e+01, 9.800e+01, 2.400e+00,
        2.640e+00, 2.800e-01, 1.370e+00, 3.700e+00, 1.180e+00, 2.690e+00,
        1.020e+03]])

In [0]:
y_train [:5]

array([2, 2, 1, 2, 0])

In [0]:
from sklearn.neighbors import KNeighborsClassifier
# эвклидова метрика
model = KNeighborsClassifier(n_neighbors=67, metric='euclidean') #'euclidean'
model.fit(X_train,y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='euclidean',
                     metric_params=None, n_jobs=None, n_neighbors=67, p=2,
                     weights='uniform')

In [0]:
y_pred = model.predict(X_test)
len(y_pred)



36

In [0]:
len(y_test)

36

In [0]:
# Анализ качества предикт accuracy_score - точность

from sklearn.metrics import accuracy_score, balanced_accuracy_score

# Выводим точность

print('Accuracy: ',accuracy_score(y_test, y_pred), accuracy_score(y_test, y_pred, normalize=False))

Accuracy:  0.8333333333333334 30


**Вывод**

Для эвклидовой метрики наилучший показатель n_neighbors=67 выдал Accuracy 0.833

**Метрика манхэттен**

In [0]:
model = KNeighborsClassifier(n_neighbors=21, metric='manhattan') #'manhattan'
model.fit(X_train,y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='manhattan',
                     metric_params=None, n_jobs=None, n_neighbors=21, p=2,
                     weights='uniform')

In [0]:
y_pred = model.predict(X_test)

In [0]:
print('Accuracy: ',accuracy_score(y_test, y_pred), accuracy_score(y_test, y_pred, normalize=False))

Accuracy:  0.7777777777777778 28


**Вывод**

n = 67, acc = 0.8333 ;
n = 50, acc = 0.8333 ;
n = 80, acc = 0.7500 ;
n = 27, acc = 0.8055 ;
n = 100, acc = 0.7777; 
n = 21, acc = 0.7777

Наилучший показатель n  = 67

